hazelcast/src/main/java/com/hazelcast/sql/impl/expression/string/SubstringFunction.java

```diff
@@ -0,0 +1,135 @@
+import com.hazelcast.sql.impl.expression.TriExpression;
+import com.hazelcast.sql.impl.expression.math.MathFunctionUtils;
+import com.hazelcast.sql.impl.row.Row;
+import com.hazelcast.sql.impl.type.QueryDataType;
+
+public class SubstringFunction extends TriExpression<String> implements IdentifiedDataSerializable {
+    public SubstringFunction() {
+        // No-op
+    }
+
+    private SubstringFunction(Expression<?> input, Expression<?> start, Expression<?> length) {
+        super(input, start, length);
+    }
+
+    public static SubstringFunction create(Expression<?> input, Expression<?> start, Expression<?> length) {
+        return new SubstringFunction(input, start, length);
+    }
+
+    @SuppressWarnings({"checkstyle:CyclomaticComplexity", "checkstyle:NPathComplexity"})
+    @Override
+    public String eval(Row row, ExpressionEvalContext context) {
+        // Get input
+        String input;
+
+        try {
+            input = StringExpressionUtils.asVarchar(operand1, row, context);
+        } catch (Exception e) {
+            // Conversion to String failed. E.g. NPE on UserClass.toString()
+            throw QueryException.dataException("Failed to get value of input operand of SUBSTRING function", e);
+        }
+
+        if (input == null) {
+            // NULL always yields NULL
+            return null;
+        }
+
+        Integer start;
+
+        if (operand2 != null) {
+            start = MathFunctionUtils.asInt(operand2, row, context);
+
+            if (start == null) {
+                return null;
+            }
+        } else {
+            start = null;
+        }
+
+        if (start == null) {
+            // Start position is not specific, start with the very first character
+            start = 1;
+        } else if (start < 1) {
+            // Different databases provide different semantics on negative values. Oracle start counting
+            // from the end, SQL Server starts from the beginning, and uses the value to calculate the
+            // final length, etc.
+            // These semantics are pretty complicated, so wi disallow it completely.
+            throw QueryException.dataException("SUBSTRING \"start\" operand must be positive");
+        }
+
+        // In SQL start position is 1-based. Convert it to 0-based for Java.
+        int adjustedStart = start - 1;
+
+        if (adjustedStart > input.length()) {
+            // Start position is beyond the string length, e.g. SUBSTRING("abc", 4)
+            return "";
+        }
+
+        Integer length;
+
+        if (operand3 != null) {
+            length = MathFunctionUtils.asInt(operand3, row, context);
+
+            if (length == null) {
+                return null;
+            }
+        } else {
+            length = null;
+        }
+
+        if (length == null) {
+            // Length is not specified, just cut from the start
+            return adjustedStart > 0 ? input.substring(adjustedStart) : input;
```

Comment:
The `start == 0` case is handled again in `String.substring`:
```suggestion
            return input.substring(adjustedStart);
```
Again on line 114

Reply:
I didn't get the comment. Could you please clarify the problem? 
If this is about two branches with identical content (`if (length == null)` and `if (adjustedStart + length > input.length())`), then this is intentional to make code simpler to understand

Reply:
The point was that we don't to special-case when `adjustedStart == 0`. `input.substring(0)` works just fine, it returns `input`. The same test for `start == 0` is inside of `String.substring`.

Reply:
I see. Formally, `String.substring()` still does additional checks, so I'd better avoid calling it as a micro-opt